src/main/java/build/buildfarm/worker/UploadManifest.java

```diff
@@ -0,0 +1,338 @@
+    tree.setRoot(root);
+
+    ByteString blob = tree.build().toByteString();
+    Digest digest = digestUtil.compute(blob);
+    Chunker chunker = new Chunker(blob, digest);
+
+    if (result != null) {
+      result
+          .addOutputDirectoriesBuilder()
+          .setPath(execRoot.relativize(dir).toString())
+          .setTreeDigest(digest);
+    }
+
+    digestToChunkers.put(chunker.digest(), chunker);
+  }
+
+  private LinkOption[] linkOpts(boolean followSymlinks) {
+    return followSymlinks ? NO_LINK_OPTION : NOFOLLOW_LINKS_OPTION;
+  }
+
+  /**
+   * Returns the status of a file.
+   */
+  private FileStatus stat(final Path path, final boolean followSymlinks) throws IOException {
+    final BasicFileAttributes attributes;
+    try {
+      attributes =
+          Files.readAttributes(path, BasicFileAttributes.class, linkOpts(followSymlinks));
+    } catch (java.nio.file.FileSystemException e) {
+      throw new FileNotFoundException(path + ERR_NO_SUCH_FILE_OR_DIR);
+    }
+    FileStatus status = new FileStatus() {
+      @Override
+      public boolean isFile() {
+        return attributes.isRegularFile() || isSpecialFile();
+      }
+
+      @Override
+      public boolean isSpecialFile() {
+        return attributes.isOther();
+      }
+
+      @Override
+      public boolean isDirectory() {
+        return attributes.isDirectory();
+      }
+
+      @Override
+      public boolean isSymbolicLink() {
+        return attributes.isSymbolicLink();
+      }
+
+      @Override
+      public long getSize() throws IOException {
+        return attributes.size();
+      }
+
+      @Override
+      public long getLastModifiedTime() throws IOException {
+        return attributes.lastModifiedTime().toMillis();
+      }
+
+      @Override
+      public long getLastChangeTime() {
+        // This is the best we can do with Java NIO...
+        return attributes.lastModifiedTime().toMillis();
+      }
+
+      @Override
+      public long getNodeId() {
+        // TODO(bazel-team): Consider making use of attributes.fileKey().
+        return -1;
+      }
+    };
+
+    return status;
+  }
+
+  /**
+   * Like stat(), but returns null on failures instead of throwing.
+   */
+  private FileStatus statNullable(Path path, boolean followSymlinks) {
+    try {
+      return stat(path, followSymlinks);
+    } catch (IOException e) {
+      return null;
+    }
+  }
+
+  private FileStatus statIfFound(Path path, boolean followSymlinks) {
+    try {
+      return stat(path, followSymlinks);
+    } catch (FileNotFoundException e) {
+      return null;
+    } catch (IOException e) {
+      // If this codepath is ever hit, then this method should be rewritten to properly distinguish
+      // between not-found exceptions and others.
+      throw new IllegalStateException(e);
+    }
+  }
+
+  private static Dirent.Type direntTypeFromStat(FileStatus stat) {
+    if (stat == null) {
+      return Dirent.Type.UNKNOWN;
+    } else if (stat.isSpecialFile()) {
```

Comment:
nit: remove all the else's:

     if (<case1>) {
       return <v1>
     }
     if (<case2>) {
       return <v2>
     }

etc